File: src/api/dcps/sacs/code/DDS/WaitSet.cs
/*
 *                         OpenSplice DDS
 *
 *   This software and documentation are Copyright 2006 to TO_YEAR PrismTech
 *   Limited, its affiliated companies and licensors. All rights reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Diagnostics;
using DDS;
using DDS.OpenSplice;
using DDS.OpenSplice.OS;
using DDS.OpenSplice.Kernel;
using DDS.OpenSplice.User;
using DDS.OpenSplice.Database;
using DDS.OpenSplice.CustomMarshalers;

namespace DDS
{
    internal struct WaitActionArg
    {
        internal int nrTriggeredConditions;
        internal int maxConditions;
        internal ICondition[] triggeredConditions;
        internal IGuardCondition[] attachedGuards;

        internal WaitActionArg(ICondition[] preAllocated)
        {
            nrTriggeredConditions = 0;
            maxConditions = 0;
            triggeredConditions = preAllocated;
            attachedGuards = null;
        }

        internal void Add(ICondition cond)
        {
            /*
             * Try to keep the amount of allocations minimal, so assume the preallocated
             * array is big enough. If that is not the case, then allocate the worst-case
             * size, so that there is never more than 1 reallocation.
             */
            ++nrTriggeredConditions;
            if ((triggeredConditions == null) || (triggeredConditions.Length < nrTriggeredConditions))
            {
                ICondition[] tmpCondArr = new ICondition[maxConditions];
                if (triggeredConditions != null) {
                    for (int i = 0; i < triggeredConditions.Length; i++)
                    {
                        tmpCondArr[i] = triggeredConditions[i];
                    }
                }
                triggeredConditions = tmpCondArr;
            }
            triggeredConditions[nrTriggeredConditions - 1] = cond;
        }

        internal ICondition[] Finalize()
        {
            /*
             * Make sure the array that is to be returned is not too big. Replace with a smaller
             * array when needed.
             */
            if (triggeredConditions != null && nrTriggeredConditions != triggeredConditions.Length)
            {
                ICondition[] tmpCondArr = new ICondition[nrTriggeredConditions];
                for(int i = 0; i < nrTriggeredConditions; i++)
                {
                    tmpCondArr[i] = triggeredConditions[i];
                }
                triggeredConditions = tmpCondArr;
            }
            return triggeredConditions;
        }
    }

    /// <summary>
    /// A WaitSet object allows an application to wait until one or more of the attached
    /// Condition objects evaluates to true or until the timeout expires.
    /// The WaitSet has no factory and must be created by the application. It is directly
    /// created as an object by using WaitSet constructors.
    /// </summary>
    public class WaitSet : SacsSuperClass, IWaitSet
    {
        private List<ICondition> conditionList = new List<ICondition>();
        private List<IGuardCondition> guardList = new List<IGuardCondition>();

        /// <summary>
        /// The default WaitSet constructor. Creates a WaitSet object.
        /// </summary>
        public WaitSet()
        {
            init();
        }

        internal ReturnCode init()
        {
            ReturnCode result;
            IntPtr uWaitSet = DDS.OpenSplice.User.WaitSet.New2();

            if (uWaitSet != IntPtr.Zero)
            {
                result = base.init(uWaitSet, true);
            }
            else
            {
                result = DDS.ReturnCode.OutOfResources;
                ReportStack.Report(result, "u_waitsetNew returned NULL");
                Environment.Exit(-1);
            }
            return result;
        }

        internal override ReturnCode wlReq_deinit()
        {
            ReturnCode result = DDS.ReturnCode.Ok;

            // In normal circumstances, no more Conditions can be attached to the WaitSet at
            // this point in time, since each attached condition has a ref to the WaitSet,
            // avoiding it from becoming garbage collected.
            // However, when an exit handler terminates the application or the application
            // runs out of its main, then both the waitset and its conditions may still be
            // alive, and so both their destructors may be called in random order.
            // Because of that scenario, we walk over the condition list and remove all
            // atached conditions.
            DDS.OpenSplice.User.WaitSet.AnnounceDestruction(rlReq_UserPeer);
            ICondition[] condArr = conditionList.ToArray();
            foreach(ICondition cond in condArr)
            {
                Condition condImpl = cond as Condition;
                result = condImpl.DetachFromWaitSet(this);
                if (result != DDS.ReturnCode.Ok) break;
            }
            if (result == DDS.ReturnCode.Ok)
            {
                IGuardCondition[] guardArr = guardList.ToArray();
                foreach(IGuardCondition guard in guardArr)
                {
                    GuardCondition guardImpl = guard as GuardCondition;
                    result = guardImpl.DetachFromWaitSet(this);
                    if (result != DDS.ReturnCode.Ok) break;
                }
            }
            if (result == DDS.ReturnCode.Ok)
            {
                result = base.wlReq_deinit();
            }
            return result;
        }

        internal int WaitAction(IntPtr userData, IntPtr arg)
        {
            int proceed = 1;

            // Extract the condition list from the pointer and add the condition to it.
            GCHandle argGCHandle = GCHandle.FromIntPtr(arg);
            WaitActionArg actionArg = (WaitActionArg) argGCHandle.Target;

            if (userData == IntPtr.Zero)
            {
                foreach (IGuardCondition guardCond in actionArg.attachedGuards)
                {
                    if (guardCond.GetTriggerValue())
                    {
                        actionArg.Add(guardCond);
                    }
                }
                proceed = (actionArg.nrTriggeredConditions == 0) ? 1 : 0;
            }
            else
            {
                GCHandle condGCHandle = GCHandle.FromIntPtr(userData);
                ICondition cond = condGCHandle.Target as ICondition;
                actionArg.Add(cond);
            }
            argGCHandle.Target = actionArg;
            return proceed;
        }

        /// <summary>
        /// This operation allows an application thread to wait for the occurrence of at least one
        /// of the conditions that is attached to the WaitSet.
        /// </summary>
        /// <param name="activeConditions">a sequence which is used to pass the list of all the attached
        /// conditions that have a trigger_value of true.</param>
        /// <param name="timeout">the maximum duration to block for the wait, after which the application thread
        /// is unblocked. The special constant Infinite can be used when the maximum waiting time does not
        /// need to be bounded.</param>
        /// <returns>Possible return codes for the operation are:
        /// Ok, Error, OutOfResources, Timeout or PreconditionNotMet</returns>
        public ReturnCode Wait(ref ICondition[] activeConditions, Duration timeout)
        {
            ReturnCode result = DDS.ReturnCode.Ok;

            ReportStack.Start();
            WaitActionArg arg = new WaitActionArg(activeConditions);

            if (QosManager.countErrors(timeout) > 0)
            {
                result = DDS.ReturnCode.BadParameter;
                ReportStack.Report (result, "Duration timeout incorrect");
            }

            while (result == DDS.ReturnCode.Ok && arg.nrTriggeredConditions == 0)
            {
                lock(this)
                {
                    arg.maxConditions = conditionList.Count + guardList.Count;
                    arg.attachedGuards = guardList.ToArray();
                    if (activeConditions == null) activeConditions = new ICondition[arg.maxConditions];
                }

                GCHandle argGCHandle = GCHandle.Alloc(arg, GCHandleType.Normal);

                V_RESULT uResult = DDS.OpenSplice.User.WaitSet.WaitAction2(
                                       rlReq_UserPeer,
                                       WaitAction,
                                       GCHandle.ToIntPtr(argGCHandle),
                                       timeout.OsDuration);

                if (uResult == V_RESULT.DETACHING)
                {
                    arg = (WaitActionArg) argGCHandle.Target;
                    lock(this)
                    {
                        foreach(ICondition cond in conditionList)
                        {
                            Condition condImpl = cond as Condition;
                            if (condImpl.IsAlive() == DDS.ReturnCode.AlreadyDeleted)
                            {
                                arg.Add(cond);
                            }
                        }
                    }
                    result = DDS.ReturnCode.Ok;
                } else {
                    result = SacsSuperClass.uResultToReturnCode(uResult);
                    arg = (WaitActionArg) argGCHandle.Target;
                }

                argGCHandle.Free();

                activeConditions = arg.Finalize();
            }
            ReportStack.Flush(this, (result != ReturnCode.Ok ) && (result != ReturnCode.Timeout));

            return result;
        }

        /**
         * Condition specific callback from non-GuardConditions to WaitSet.
         */
        internal ReturnCode wlReq_AttachGeneralCondition (ICondition condition, IntPtr uCondition, IntPtr condGCHandle)
        {
            ReturnCode result;

            Debug.Assert (!conditionList.Contains (condition));
            result = SacsSuperClass.uResultToReturnCode (
                    DDS.OpenSplice.User.WaitSet.Attach (
                            rlReq_UserPeer, uCondition, condGCHandle));
            if (result == DDS.ReturnCode.Ok) {
                conditionList.Add (condition);
                MyDomainId = DDS.OpenSplice.User.WaitSet.GetDomainId(rlReq_UserPeer);
            } else {
                ReportStack.Report (result, "Could not attach Condition to WaitSet.");
            }


            return result;
        }

        /**
         * Condition specific callback from GuardConditions to WaitSet.
         */
        internal ReturnCode wlReq_AttachGuardCondition(IGuardCondition guardCond)
        {
            ReturnCode result;

            Debug.Assert(!guardList.Contains(guardCond));
            result = SacsSuperClass.uResultToReturnCode(
                    DDS.OpenSplice.User.WaitSet.Notify(rlReq_UserPeer, IntPtr.Zero));
            if (result == DDS.ReturnCode.Ok)
            {
                guardList.Add(guardCond);
            } else {
                ReportStack.Report (result, "Could not attach GuardCondition to WaitSet.");
            }
            return result;
        }

        /**
         * Condition specific callback from non-GuardConditions to WaitSet.
         */
        internal ReturnCode wlReq_DetachGeneralCondition(ICondition condition, IntPtr uCondition)
        {
            ReturnCode result;

            Debug.Assert(conditionList.Contains(condition));
            result = SacsSuperClass.uResultToReturnCode(
                    DDS.OpenSplice.User.WaitSet.Detach(
                            rlReq_UserPeer, uCondition));
            if (result == DDS.ReturnCode.Ok)
            {
                conditionList.Remove(condition);
                MyDomainId = DDS.OpenSplice.User.WaitSet.GetDomainId(rlReq_UserPeer);
            } else {
                ReportStack.Report (result, "Could not detach Condition from WaitSet.");
            }
            return result;
        }

        /**
         * Condition specific callback from GuardConditions to WaitSet.
         */
        internal ReturnCode wlReq_DetachGuardCondition(IGuardCondition guardCond)
        {
            ReturnCode result;

            Debug.Assert(guardList.Contains(guardCond));
            result = SacsSuperClass.uResultToReturnCode(
                    DDS.OpenSplice.User.WaitSet.Notify(rlReq_UserPeer, IntPtr.Zero));
            if (result == DDS.ReturnCode.Ok)
            {
               guardList.Remove(guardCond);
            } else {
                ReportStack.Report (result, "Could not detach GuardCondition from WaitSet.");
            }
            return result;
        }

        /// <summary>
        /// This operation attaches a condition to the WaitSet
        /// </summary>
        /// <param name="condition">The condition to be attached to the WaitSet.
        /// The parameter must be either a ReadCondition, QueryCondition, StatusCondition or GuardCondition</param>
        /// <returns>Possible return codes are: Ok,Error,BadParameter or OutOfResources.</returns>
        public ReturnCode AttachCondition (ICondition condition)
        {
            ReturnCode result = DDS.ReturnCode.BadParameter;
            Condition condImpl = condition as Condition;
            ReportStack.Start ();

            if (condImpl == null) {
                ReportStack.Report (result, "cond is invalid (null), or not of type " +
                                            "DDS::OpenSplice::Condition");
            } else {
                lock(this)
                {
                    result = condImpl.AttachToWaitSet (this);

                    // ALREADY_DELETED may only apply to the WaitSet in this context,
                    // so for a deleted condition use BAD_PARAMETER instead.
                    if (result == DDS.ReturnCode.AlreadyDeleted) {
                        result = DDS.ReturnCode.BadParameter;
                    }
                }
            }
            ReportStack.Flush(this, result != ReturnCode.Ok);

            return result;
        }

        /// <summary>
        /// This operation detaches a Condition from the WaitSet. If the Condition was not attached
        /// to this WaitSet, the operation returns PreconditionNotMet
        /// </summary>
        /// <param name="condition">The attached condition in the WaitSet which is to be detached.</param>
        /// <returns>Possible return codes are: Ok, Error, BadParameter, OutOfResources or PreconditionNotMet.</returns>
        public ReturnCode DetachCondition (ICondition condition)
        {
            ReturnCode result = DDS.ReturnCode.BadParameter;
            Condition condImpl = condition as Condition;
            ReportStack.Start ();

            if (condImpl == null) {
                ReportStack.Report (result, "cond is invalid (null), or not of type " +
                                    "DDS::OpenSplice::Condition");
            } else {
                lock(this)
                {
                    result = condImpl.DetachFromWaitSet (this);

                    // ALREADY_DELETED may only apply to the WaitSet in this context,
                    // so for a deleted condition use BAD_PARAMETER instead.
                    if (result == DDS.ReturnCode.AlreadyDeleted) {
                        result = DDS.ReturnCode.BadParameter;
                    }
                }
            }
            ReportStack.Flush(this, result != ReturnCode.Ok);
            return result;
        }

        /// <summary>
        /// This operation retrieves the list of attached conditions.
        /// </summary>
        /// <param name="attachedConditions">A reference to a sequence that will hold all the attached conditions
        /// on the WaitSet</param>
        /// <returns>Possible return codes are: Ok, Error or OutOfResources.</returns>
        public ReturnCode GetConditions(ref ICondition[] attachedConditions)
        {
            lock(this)
            {
                int i = 0;
                int total = conditionList.Count + guardList.Count;
                if (attachedConditions == null || attachedConditions.Length != total)
                {
                    attachedConditions = new ICondition[total];
                }
                foreach(ICondition cond in conditionList)
                {
                    attachedConditions[i++] = cond;
                }
                foreach(ICondition cond in guardList)
                {
                    attachedConditions[i++] = cond;
                }
            }

            return DDS.ReturnCode.Ok;
        }

        internal ReturnCode trigger (IntPtr context)
        {
            ReturnCode result;

            lock(this)
            {
                result = SacsSuperClass.uResultToReturnCode (
                        DDS.OpenSplice.User.WaitSet.Notify (rlReq_UserPeer, context));
                if (result != ReturnCode.Ok) {
                    ReportStack.Report (result, "WaitSet notify failed ");
                }
            }
            return result;
        }
    }
}

File: src/api/dcps/c++/common/code/WaitSet.cpp
/*
 *                         OpenSplice DDS
 *
 *   This software and documentation are Copyright 2006 to TO_YEAR PrismTech
 *   Limited, its affiliated companies and licensors. All rights reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
#include "u_user.h"
#include "u_observable.h"
#include "c_base.h"
#include "WaitSet.h"
#include "GuardCondition.h"
#include "StatusCondition.h"
#include "ReadCondition.h"
#include "ReportUtils.h"

DDS::WaitSet_ptr
DDS::WaitSet::_nil (void)
{
    return static_cast<DDS::WaitSet_ptr>(NULL);
}

DDS::WaitSet_ptr
DDS::WaitSet::_duplicate (DDS::WaitSet_ptr obj)
{
    return dynamic_cast<DDS::WaitSet_ptr>(
        DDS::WaitSetInterface::_duplicate (obj));
}

DDS::WaitSet_ptr
DDS::WaitSet::_narrow (DDS::WaitSet_ptr obj)
{
    return dynamic_cast<DDS::WaitSet_ptr>(
        DDS::WaitSetInterface::_narrow (obj));
}

DDS::WaitSet_ptr
DDS::WaitSet::_unchecked_narrow (DDS::WaitSet_ptr obj)
{
    return dynamic_cast<DDS::WaitSet_ptr>(
        DDS::WaitSetInterface::_narrow (obj)); /* _unchecked_narrow is not implemented by all ORBs. */
}

DDS::WaitSet::WaitSet (
) : DDS::OpenSplice::CppSuperClass(DDS::OpenSplice::WAITSET),
    uWaitset(NULL),
    conditions(new DDS::OpenSplice::ObjSet(FALSE)),
    guards(new DDS::OpenSplice::ObjSet(TRUE))
{
    (void) init();
}

DDS::WaitSet::~WaitSet ()
{
    (void) deinit();
    delete conditions;
    delete guards;
}

DDS::ReturnCode_t
DDS::WaitSet::init()
{
    return nlReq_init();
}

DDS::ReturnCode_t
DDS::WaitSet::nlReq_init()
{
    DDS::ReturnCode_t result;

    result = conditions->init();
    if (result == DDS::RETCODE_OK) {
        result = guards->init();
    }
    if (result == DDS::RETCODE_OK) {
        result = DDS::OpenSplice::CppSuperClass::nlReq_init();
    }
    if (result == DDS::RETCODE_OK) {
        this->uWaitset = u_waitsetNew2 ();
        if (this->uWaitset == NULL) {
            result = DDS::RETCODE_OUT_OF_RESOURCES;
            CPP_PANIC("Could not create WaitSet.");
        }
    }
    return result;
}

DDS::ReturnCode_t
DDS::WaitSet::wlReq_deinit()
{
    DDS::ReturnCode_t result = DDS::RETCODE_OK;
    DDS::ObjSeq_var condSeq = conditions->getObjSeq();
    DDS::ULong length = condSeq->length();

    result = DDS::OpenSplice::CppSuperClass::wlReq_deinit();
    if (result == DDS::RETCODE_OK) {
        u_waitsetAnnounceDestruction(uWaitset);
        for (DDS::ULong i = 0; i < length && result == DDS::RETCODE_OK; i++) {
            DDS::OpenSplice::Condition *cond;

            cond = dynamic_cast<DDS::OpenSplice::Condition *>(condSeq[i].in());
            result = cond->detachFromWaitset(this);
        }
    }

    if (result == DDS::RETCODE_OK) {
        condSeq = guards->getObjSeq();
        length = condSeq->length();
        for (DDS::ULong i = 0; i < length && result == DDS::RETCODE_OK; i++) {
            DDS::OpenSplice::Condition *cond;

            cond = dynamic_cast<DDS::OpenSplice::Condition *>(condSeq[i].in());
            result = cond->detachFromWaitset(this);
        }
        u_waitsetNotify (uWaitset, NULL);
        result = uResultToReturnCode(u_objectFree_s (u_object (uWaitset)));
        if (result == DDS::RETCODE_OK) {
            uWaitset = NULL;
            conditions->clear();
            guards->clear();
            result = conditions->deinit();
        }
        if (result == DDS::RETCODE_OK) {
            result = guards->deinit();
        }
    }

    return result;
}

DDS::Boolean
DDS::WaitSet::collect_detached_conditions(
    DDS::Object_ptr td,
    struct findObjectArg *arg)
{
    DDS::ConditionSeq *activeConditions = reinterpret_cast<DDS::ConditionSeq *>(arg);
    DDS::ULong length = activeConditions->length();
    DDS::OpenSplice::Condition *condition = dynamic_cast<DDS::OpenSplice::Condition *>(td);
    assert(condition); /* Type already checked during insertion. */

    if (condition->isAlive() == DDS::RETCODE_ALREADY_DELETED) {
        activeConditions->length(++length);
        (*activeConditions)[length - 1] = DDS::Condition::_duplicate(condition);
    }

    return TRUE;
}

typedef struct {
    DDS::ConditionSeq *activeConditions;// Don't clean up afterwards: this field is returned to the caller!!
    DDS::ObjSeq_var guards;             // _var type automatically cleans up afterward.
    DDS::Long maxConditions;
} WaitActionArg;

static os_boolean
wait_action (
    c_voidp userData,
    c_voidp arg)
{
    DDS::Condition_ptr condition;
    os_boolean proceed = OS_TRUE;
    WaitActionArg *waitArg = reinterpret_cast<WaitActionArg *>(arg);
    DDS::ConditionSeq &condSeq = *(waitArg->activeConditions);
    DDS::ULong length = condSeq.length();
    DDS::ULong max = condSeq.maximum();

    if (userData == NULL) {
        DDS::ULong nrGuards = waitArg->guards->length();
        for (DDS::ULong i = 0; i < nrGuards; i++)
        {
            condition = dynamic_cast<DDS::Condition_ptr>(waitArg->guards[i].in());
            assert(condition);
            if (condition->get_trigger_value()) {
                if (++length > max)
                {
                    // Stretch sequence up to worst-case, to avoid continuous re-allocations.
                    max = waitArg->maxConditions;
                    condSeq.length(max);
                }
                condSeq.length(length);
                condSeq[length - 1] = DDS::Condition::_duplicate(condition);
            }
        }
        proceed = (length == 0) ? OS_TRUE : OS_FALSE;
    } else {

        condition = reinterpret_cast<DDS::Condition_ptr>(userData);
        assert(condition);
        condSeq.length(++length);
        condSeq[length - 1] = DDS::Condition::_duplicate(condition);
    }
    return proceed;
}

DDS::ReturnCode_t
DDS::WaitSet::wait(
    ConditionSeq &active_conditions,
    const DDS::Duration_t &timeout
) THROW_ORB_EXCEPTIONS
{
    DDS::ReturnCode_t result = DDS::RETCODE_OK;
    os_duration uTimeout;
    WaitActionArg arg;

    CPP_REPORT_STACK();

    active_conditions.length(0);
    arg.activeConditions = &active_conditions;
    arg.guards = NULL;

    result = DDS::OpenSplice::Utils::copyDurationIn(timeout, uTimeout);

    while (result == DDS::RETCODE_OK && active_conditions.length() == 0) {
        result = this->read_lock ();
        if (result == DDS::RETCODE_OK) {
            arg.maxConditions = conditions->getNrElements() + guards->getNrElements();
            /* Avoid re-allocating a new sequence every iteration, so only create it the first time. */
            if (arg.guards == NULL) {
                arg.guards = guards->getObjSeq();
            } else {
                guards->getObjSeq(arg.guards);
            }
            this->unlock ();

            u_result uResult = u_waitsetWaitAction2(uWaitset, wait_action, &arg, uTimeout);
            if (uResult == U_RESULT_DETACHING) {
                result = this->read_lock ();
                if(result == DDS::RETCODE_OK) {
                    (void) this->conditions->walk(
                            (DDS::OpenSplice::ObjSet::ObjSetActionFunc)collect_detached_conditions,
                            &active_conditions);
                    this->unlock ();
                }
            } else {
                result = uResultToReturnCode(uResult);
            }
        }
    }

    CPP_REPORT_FLUSH(this, (result != DDS::RETCODE_OK) && (result != DDS::RETCODE_TIMEOUT));

    return result;
}

/**
 * Condition specific callback from non-GuardConditions to WaitSet.
 */
DDS::ReturnCode_t
DDS::WaitSet::wlReq_attachGeneralCondition(
    DDS::Condition_ptr condition,
    u_observable uCondition)
{
    DDS::ReturnCode_t result;
    u_result uResult;

    /* The WaitSet does not modify the reCounts of the conditions it points to: the destructor of
     * each condition is supposed to detach itself from the WaitSet. Keeping track of refCounts would
     * avoid some Conditions to ever reach 0 due to cyclic references.
     */
    assert (this->conditions->containsElement(condition) == FALSE);

    uResult = u_waitsetAttach(uWaitset, uCondition, condition);
    result = uResultToReturnCode(uResult);
    if (result == DDS::RETCODE_OK) {
        DDS::Boolean inserted = this->conditions->insertElement(condition);
        assert (inserted);
        OS_UNUSED_ARG(inserted);
        this->setDomainId(u_waitsetGetDomainId(this->uWaitset));
    } else {
        CPP_REPORT(result, "Could not attach Condition to WaitSet.");
    }

    return result;
}

/**
 * Condition specific callback from GuardConditions to WaitSet.
 */
DDS::ReturnCode_t
DDS::WaitSet::wlReq_attachGuardCondition(
    GuardCondition *condition)
{
    DDS::ReturnCode_t result;
    u_result uResult;

    /* The WaitSet does not modify the reCounts of the conditions it points to: the destructor of
     * each condition is supposed to detach itself from the WaitSet. Keeping track of refCounts would
     * avoid some Conditions to ever reach 0 due to cyclic references.
     */
    assert (this->guards->containsElement(condition) == FALSE);

    uResult = u_waitsetNotify(uWaitset, NULL);
    result = uResultToReturnCode(uResult);
    if (result == DDS::RETCODE_OK) {
        DDS::Boolean inserted = this->guards->insertElement(condition);
        assert (inserted);
        OS_UNUSED_ARG(inserted);
    } else {
        CPP_REPORT(result, "Could not attach GuardCondition to WaitSet.");
    }

    return result;
}

/**
 * Condition specific callback from non-GuardConditions to WaitSet.
 */
DDS::ReturnCode_t
DDS::WaitSet::wlReq_detachGeneralCondition(
    DDS::OpenSplice::Condition *condition,
    u_observable uCondition)
{
    DDS::ReturnCode_t result;
    u_result uResult;

    /* The WaitSet does not modify the reCounts of the conditions it points to: the destructor of
     * each condition is supposed to detach itself from the WaitSet. Keeping track of refCounts would
     * avoid some Conditions to ever reach 0 due to cyclic references.
     */
    assert (this->conditions->containsElement(condition));

    uResult = u_waitsetDetach_s(uWaitset, uCondition);
    result = uResultToReturnCode(uResult);
    if ((result == DDS::RETCODE_OK) || (result == DDS::RETCODE_ALREADY_DELETED)) {
        // Set takes care of its own refCounting.
        DDS::Boolean removed = this->conditions->removeElement(condition);
        assert (removed);
        OS_UNUSED_ARG(removed);
        this->setDomainId(u_waitsetGetDomainId(this->uWaitset));
    } else {
        CPP_REPORT(result, "Could not detach Condition from WaitSet.");
    }

    return result;
}

/**
 * Condition specific callback from GuardConditions to WaitSet.
 */
DDS::ReturnCode_t
DDS::WaitSet::wlReq_detachGuardCondition(
    GuardCondition *condition)
{
    DDS::ReturnCode_t result;
    u_result uResult;

    /* The WaitSet does not modify the reCounts of the conditions it points to: the destructor of
     * each condition is supposed to detach itself from the WaitSet. Keeping track of refCounts would
     * avoid some Conditions to ever reach 0 due to cyclic references.
     */
    assert (this->guards->containsElement(condition));

    uResult = u_waitsetNotify(uWaitset, NULL);
    result = uResultToReturnCode(uResult);
    if (result == DDS::RETCODE_OK) {
        // Set takes care of its own refCounting.
        DDS::Boolean removed = this->guards->removeElement(condition);
        assert (removed);
        OS_UNUSED_ARG(removed);
    } else {
        CPP_REPORT(result, "Could not detach GuardCondition from WaitSet.");
    }

    return result;
}

DDS::ReturnCode_t
DDS::WaitSet::attach_condition(
    DDS::Condition_ptr cond
) THROW_ORB_EXCEPTIONS
{
    DDS::ReturnCode_t result = DDS::RETCODE_OK;
    DDS::OpenSplice::Condition *condition;

    CPP_REPORT_STACK();

    if (cond == NULL) {
        result = DDS::RETCODE_BAD_PARAMETER;
        CPP_REPORT(result, "cond '<NULL>' is invalid.");
    } else {
        condition = dynamic_cast<DDS::OpenSplice::Condition *>(cond);
        if (condition == NULL) {
            result = DDS::RETCODE_BAD_PARAMETER;
            CPP_REPORT(result, "cond is invalid, not of type '%s'.",
                "DDS::OpenSplice::Condition");
        }
    }

    if (result == DDS::RETCODE_OK) {
        result = this->write_lock();
        if (result == DDS::RETCODE_OK) {
            result = condition->attachToWaitset(this);

            // ALREADY_DELETED may only apply to the WaitSet in this context,
            // so for a deleted condition use BAD_PARAMETER instead.
            if (result == DDS::RETCODE_ALREADY_DELETED) {
                result = DDS::RETCODE_BAD_PARAMETER;
            }
            this->unlock();
        }
    }

    CPP_REPORT_FLUSH(this, result != DDS::RETCODE_OK);

    return result;
}

DDS::ReturnCode_t
DDS::WaitSet::detach_condition(
    DDS::Condition_ptr cond
) THROW_ORB_EXCEPTIONS
{
    DDS::ReturnCode_t result = DDS::RETCODE_OK;
    DDS::OpenSplice::Condition *condition;

    CPP_REPORT_STACK();

    if (cond == NULL) {
        result = DDS::RETCODE_BAD_PARAMETER;
        CPP_REPORT(result, "cond '<NULL>' is invalid.");
    } else {
        condition = dynamic_cast<DDS::OpenSplice::Condition *>(cond);
        if (condition == NULL) {
            result = DDS::RETCODE_BAD_PARAMETER;
            CPP_REPORT(result, "cond is invalid, not of type '%s'.",
                "DDS::OpenSplice::Condition");
        }
    }

    if (result == DDS::RETCODE_OK) {
        result = this->write_lock ();
        if (result == DDS::RETCODE_OK) {
            result = condition->detachFromWaitset(this);

            // ALREADY_DELETED may only apply to the WaitSet in this context,
            // so for a deleted condition use BAD_PARAMETER instead.
            if (result == DDS::RETCODE_ALREADY_DELETED) {
                result = DDS::RETCODE_BAD_PARAMETER;
            }
            this->unlock ();
        }
    }

    CPP_REPORT_FLUSH(this, result != DDS::RETCODE_OK);

    return result;
}

static DDS::Boolean
copy_to_sequence (
    DDS::Object_ptr element,
    void *arg)
{
    DDS::Condition_ptr condition =
        dynamic_cast<DDS::Condition_ptr>(element);
    assert (condition != NULL);
    DDS::ConditionSeq *attached_conditions =
        reinterpret_cast<DDS::ConditionSeq *>(arg);
    assert (attached_conditions != NULL);
    DDS::ULong length;

    length = attached_conditions->length ();
    attached_conditions->length (length + 1);
    (*attached_conditions)[length] = DDS::Condition::_duplicate (condition);
    return TRUE;
}

DDS::ReturnCode_t
DDS::WaitSet::get_conditions(
    DDS::ConditionSeq &attached_conditions)
{
    DDS::ReturnCode_t result;

    CPP_REPORT_STACK();

    result = this->read_lock ();
    if (result == DDS::RETCODE_OK) {
        attached_conditions.length(conditions->getNrElements()+guards->getNrElements());
        attached_conditions.length(0);
        (void)conditions->walk(copy_to_sequence, &attached_conditions);
        guards->walk(copy_to_sequence, &attached_conditions);
        this->unlock ();
    }

    CPP_REPORT_FLUSH(this, result != DDS::RETCODE_OK);

    return result;
}

DDS::ReturnCode_t
DDS::WaitSet::trigger (
    DDS::Condition_ptr cond)
{
    DDS::ReturnCode_t result;
    u_result uResult;

    assert (cond != NULL);

    result = this->check ();
    if (result == DDS::RETCODE_OK) {
        uResult = u_waitsetNotify (uWaitset, cond);
        result = uResultToReturnCode (uResult);
        CPP_REPORT(result, "Could not trigger WaitSet.");
    }

    return result;
}

os_int32
DDS::WaitSet::getDomainId()
{
    int domainId = -1;

    if (this->check() == DDS::RETCODE_OK) {
        domainId = u_waitsetGetDomainId(this->uWaitset);
    }

    return domainId;
}
